lib/df-detection.c

```diff
@@ -0,0 +1,282 @@
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		current->addresses[current->num_read].caller = caller;
+		current->addresses[current->num_read].stack =
+		    df_save_stack(GFP_NOWAIT);
+		detect_intersection(kernel_addr);
+		current->num_read++;
+	}
+}
+
+void start_system_call(long syscall)
+{
+	if (!current->df_enable)
+		return;
+	current->syscall_num = syscall;
+	current->addresses = (struct df_address_range *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_address_range), GFP_KERNEL);
+	current->sz = current->addresses ? DF_INIT_SIZE : 0;
+	current->num_read = 0;
+	current->pairs = (struct df_pair *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
+	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
+	current->df_index = 0;
+}
+
+void end_system_call(void)
+{
+	if (!current->df_enable)
+		return;
+	if (current->pairs != NULL) {
+		if (current->df_index)
+			report();
+		kfree(current->pairs);
+		current->pairs = NULL;
+		current->df_index = 0;
+		current->df_size = 0;
+	}
+	if (current->addresses != NULL) {
+		current->num_read = 0;
+		current->sz = 0;
+		kfree(current->addresses);
+		current->addresses = NULL;
+	}
+}
+
+void report(void)
+{
+	if (!check_valid_detection())
+		return;
+	int i;
+	unsigned long *first_entries;
+	unsigned int first_nr_entries;
+	unsigned long *second_entries;
+	unsigned int second_nr_entries;
+	unsigned long first_frame = 0;
+	unsigned long second_frame = 0;
+
+	for (i = 0; i < current->df_index; i++) {
+		if (current->addresses[current->pairs[i].first].stack &&
+		    current->addresses[current->pairs[i].second].stack) {
+			first_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].first].stack,
+			    &first_entries);
+			second_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].second].stack,
+			    &second_entries);
+			int first_index =
+			    filter_stack(first_entries, first_nr_entries);
+			int second_index =
+			    filter_stack(second_entries, second_nr_entries);
+			first_frame = first_entries[first_index];
+			second_frame = second_entries[second_index];
+			char first_buf[64], second_buf[64];
+			int first_len, second_len;
+			/*check if this bug happent in perf_copy_attr*/
+			first_len = scnprintf(first_buf, sizeof(first_buf),
+					      "%ps", (void *)first_frame);
+			second_len = scnprintf(second_buf, sizeof(second_buf),
+					       "%ps", (void *)second_frame);
+			if (strnstr(first_buf, "perf_copy_attr", first_len) &&
+			    strnstr(second_buf, "perf_copy_attr", second_len))
+				continue;
+			pr_err("BUG: multi-read in %ps  "
+			       "between %ps and %ps\n ",
```

Comment:
As Dmitry noted, let's try to print only first_frame in the header without the syscall name and second_frame.
This will help to cluster similar reports together more efficiently.

Reply:
so we need 2 frames from the stack of each address range after filtering copy functions?
so by this in the title we will have 4 functions name

Reply:
and also isn't syscall name in the header make it easier to link multi read crashes with fault injection crashes

Reply:
No, what I meant is just make the above `pr_err("BUG: multi-read in %ps  ", first_frame);`

But, fair enough, preserving the syscall name may help us better link multi-read reports with crashes. Feel free to just omit second_frame then, something along the lines of: `"BUG: multi-read in %ps, syscall %ps\n "`

Reply:
yeah okay sorry i misunderstood, okay i will remove only second_frame and i will keep syscall name if we observe that it won't help us linking between multi read and fault injection, i can easily remove it